bindings/rust/s2n-tls/src/cert_chain.rs

```diff
@@ -0,0 +1,131 @@
+use crate::error::{Error, Fallible};
+use s2n_tls_sys::*;
+use std::marker::PhantomData;
+use std::ptr::{self, NonNull};
+
+/// A CertificateChain represents a chain of X.509 certificates.
+pub struct CertificateChain {
+    ptr: NonNull<s2n_cert_chain_and_key>,
+}
+
+impl CertificateChain {
+    /// This allocates a new certificate chain from s2n.
+    pub fn new() -> Result<CertificateChain, Error> {
+        unsafe {
+            let ptr = s2n_cert_chain_and_key_new().into_result()?;
+            Ok(CertificateChain { ptr })
+        }
+    }
+
+    pub fn iter(&self) -> CertificateChainIter<'_> {
+        CertificateChainIter {
+            idx: 0,
+            chain: self,
+        }
+    }
+
+    pub fn len(&self) -> Result<u32, Error> {
+        let mut length = 0;
+        unsafe {
+            s2n_cert_chain_get_length(self.ptr.as_ptr(), &mut length)
+                .into_result()
+                .map(|_| ())?;
+        }
+        Ok(length)
+    }
+
+    pub(crate) fn as_mut_ptr(&mut self) -> NonNull<s2n_cert_chain_and_key> {
+        self.ptr
+    }
+}
+
+// # Safety
+//
+// s2n_cert_chain_and_key objects can be sent across threads
+unsafe impl Send for CertificateChain {}
+
+// # Safety
+//
+// s2n_cert_chain_and_key objects can be shared across threads.
+//
+// (TODO: Verify? Connection isn't for some reason.)
```

Comment:
Seems OK. Even if this is a pointer into the Connection it's not obvious that prevents us from implementing Sync (FWIW, I didn't really find rationale for why not in the other thread, maybe we should document that somewhere) -- if the certificate itself is mostly just a bag of bytes, then it seems OK for it to be Sync even if the wider connection isn't.


> However, if you get the chain from the function s2n_connection_get_selected_cert(currently not implemented in your PR), it is a pointer to memory inside of an s2n_connection struct.

Given this might be the case, I'll go ahead and add a lifetime to CertificateChain so that we can implement that API later on.